crates/recursor/src/recursor_dns_handle.rs

```diff
@@ -24,6 +25,44 @@ use crate::{
     Error, ErrorKind,
 };
 
+/// This macro is used in ns_pool_for_zone and ns_pool_for_referral to resolve futures related to
+/// nameserver lookups.  There are a few variations of lookup methods in use, with differing future
+/// types and return types, but the resolution loop is otherwise identical, so this macro will help
+/// bridge the type differences.  This macro expects a closure to be passed (the "processor"
+/// argument,) which takes a positive future response and adds the responses to the
+/// NameServerConfigGroup in config_group.  The closure must take two arguments: an
+/// &mut NameServerConfigGroup and a response type, which should be a reference to the response type
+/// returned by the lookup future.  For instance, for resolution futures that use recursor.resolve,
+/// the response type argument should be &Lookup.  For resolution futures that use
+/// NameServerPool.lookup, the response argument should be &DnsResponse.  A typical closure looks
+/// something like this:
+///
+/// ```rust,ignore
+///     let processor = |cg: &mut NameServerConfigGroup, response: &Lookup| {
+///         cg.append_ips(response.iter().filter_map(RData::ip_addr), true);
+///     };
+/// ```
+macro_rules! process_futures {
```

Comment:
nit: at a glance I would think this macro can be written as a generic function but if that ends up being too complex, this macro is fine.

Reply:
(Did that in https://github.com/marcus0x62/hickory-dns/pull/1.)